### ISSUE TITLE
AbandonedMutexException is not being thrown when parent process dies on Unix

### ISSUE THREAD

**Issue by odhanson** [dotnet/coreclr#21455]
The issue occurs when the parent process creates and acquires a named Mutex. Then a child process is spawned which opens the named mutex and tries to acquire a lock. Then the parent process ends without releasing the mutex. According to documentation (and also verified behavior on windows) an AbandonedMutexException should be thrown in the child process. 

This works fine for sibling processes, but when one process is a parent and the other is a child, as in the following code example, this will fail: 
```c#
using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Threading;

namespace TestMutex
{
    class Program
    {
        static void Main(string[] args)
        {
            bool owner;
            owner = args.Length == 0;
            Mutex mutex = new Mutex(owner, "TestMutex");

            if (owner)
            {
                Process p = new Process();
                string codeBase = Assembly.GetEntryAssembly().CodeBase;
                Uri u = new Uri(codeBase);
                string path = Path.GetFileNameWithoutExtension(u.AbsolutePath);
                p.StartInfo.FileName = path;
                p.StartInfo.Arguments = "client";
                p.StartInfo.CreateNoWindow = true;
                p.StartInfo.UseShellExecute = false;

                Console.WriteLine($"Owner Mutex aquired. Starting process: {path}");

                p.Start();

                Console.WriteLine($"Child process created with pid: {p.Id}.");

                Thread.Sleep(10000);
                Console.WriteLine("Good bye");
            }
            else
            {
                try
                {
                    Console.WriteLine("ChildMutex Created!");
                    while (!mutex.WaitOne(1000))
                    {
                        Console.WriteLine("Waiting!");
                    }
                    Console.WriteLine("Mutex aquired.!");

                }
                catch (AbandonedMutexException)
                {
                    Console.WriteLine("Parent process ended");
                }
            }
        }
    }
}
```

You need to publish this program as self contained in order for it to work.

I have debugged the issue, and it turns out that the problem is with how we spawn new processes. We are basically using Fork and Exec in Unix systems. However, by default, the parent's opened file descriptors are kept open after fork and exec. Also, exclusive locks are [inherited by fork/exec](https://en.wikipedia.org/wiki/File_locking). So it turns out that the child process already has an exclusive lock to the mutex file (without knowing about it) and when the parent dies, this lock remains, thus we do not receive the abandoned exception.

There are multiple ways to fix this. But this depends if we actually want to inherit the fids by the child or not. For example, in the case of anonymous pipes, I assume that we do want to inherit the fids. The safest would be to mark each file we don't want to inherit with a FD_CLOEXEC (close on exec) attribute.

**Comment by odhanson** [comment]
cc: @jkotas, @janvorli, @kouvel, @jingsunsql, @danmosemsft

**Comment by danmosemsft** [comment]
@tmds

**Comment by janvorli** [comment]
Fixed by #21458
